src/CTA.FeatureDetection.ProjectType/CompiledFeatures/AspNetCoreWebApiFeature.cs

```diff
@@ -0,0 +1,33 @@
+﻿using Codelyzer.Analysis;
+using CTA.FeatureDetection.Common.Extensions;
+using CTA.FeatureDetection.Common.Models.Features.Base;
+using System.Linq;
+
+namespace CTA.FeatureDetection.ProjectType.CompiledFeatures
+{
+    public class AspNetCoreWebApiFeature : CompiledFeature
+    {
+        /// <summary>
+        /// Determines that a project is an ASP.NET Core WebAPi project if:
+        ///     1) There are any classes derived from the ControllerBase abstract class
+        ///        and no classes derived from the Controller abstract class
+        ///     OR
+        ///     2) If there are any classes derived from the Controller abstract class,
+        ///        none of them call any methods that return a view-related object.
+        /// </summary>
+        /// <param name="analyzerResult"></param>
+        /// <returns>Whether a project is an ASP.NET Core WebApi project or not</returns>
+        public override bool IsPresent(AnalyzerResult analyzerResult)
+        {
+            var project = analyzerResult.ProjectResult;
+            var controllerBaseClassDeclarations = project.GetClassDeclarationsWithBaseType(Constants.NetCoreMvcControllerBaseOriginalDefinition);
+            var controllerClassDeclarations = project.GetClassDeclarationsWithBaseType(Constants.NetCoreMvcControllerOriginalDefinition);
+            var returnsViewObjectFromControllerClasses = controllerClassDeclarations.Any(c => c.GetInvocationExpressionsBySemanticReturnType(Constants.NetCoreViewResultTypes).Any());
```

Comment:
How does this help us detect a web api? When a controller is returning a View, usually it means it's an Mvc project.

One thing we can look for is the ApiController attribute in a class declaration, although im not sure if that covers all cases. I will do some research too.

Reply:
We are looking for Views being returned because it's possible that a WebApi project will use an MVC controller. This part of the condition is: If there are any MVC Controllers, make sure they are NOT returning any Views (partial or otherwise).

I would definitely appreciate additional research here as well. I think the line between WebApi and MVC projects is much more grey in .NET Core since Microsoft merged the two libraries.

Reply:
Here's what I found:

- All projects, regardless of type (Mvc or web API) will have a non-empty views folder initially, so we can't use views as a condition
- An Mvc project can also be a web api project, and vice versa
- For MVC: Return should be a view. Not sure if you checked, but the return type of the below should be View, not IActionResult
`
public IActionResult Index()
{
    return View();
}
`
- For Web API, the condition should be: 1) Controller with ApiController attribute (regardless of what it inherits from) or 2) Controller with a public method that doesn't return a view.